### ISSUE TITLE
Call for new features 

### ISSUE THREAD

**Issue by AzHicham** [rusty-celery/rusty-celery#349]
Hello,

Two or 3 awesome feature we could have in the future:
- Task chaining !!!
- Abortable Tasks (Backend needed first I suppose) like in python celery
- Share object between tasks (cache, ..) in a generic way like in Arctix web https://docs.rs/actix-web/latest/actix_web/web/struct.Data.html

**Comment by epwalsh** [comment]
Absolutely! I think # 3 (shared object between tasks) should be high priority. I've seen that requested multiple times.

**Comment by ehiggs** [comment]
I was looking for functionality #3 as well. Without it, celery is of limited usefulness. Can't share db connections, can't share http client sessions, etc.

The pattern used by actix is common in Rust and Go web servers and is a good fit for this. I think it makes a bit more sense to investigate if this could actually be done by leveraging [Tower](https://github.com/tower-rs/tower). This way, various middlewares could be used (obviously not all relevant, like account management).

IMO actix is really amazing work but also a bit challenging to understand. Axum (the tower webserver) is by contrast simpler.

Wdyt?

**Comment by epwalsh** [comment]
I'm very familiar with Tower or Actix, but I think the simplest approach to get this done would look something like how Actix does it, as @AzHicham mentioned. In particular we'd let the user bind some arbitrary data type to the Celery App instance. Then if you want to access that from a task you'd do so via the `request` object. For example:

```rust
#[celery::task(bind = true)]
fn task_with_db_connection(task: &Self) -> TaskResult<()> {
    task.request.app.data.connection()  // or whatever
    Ok(())
}
```

**Comment by parz3val** [comment]
I would be up for it as I am using a modified version of rusty celery where I can have a state, but I am fairly new to doing rust libraries, so I don't know if my approach is good.
```
// type of state is Option<Arc<Mutex<T>>> => This so I can intialize state inside a task macro with None
#[celery:::task(bind = true) ]
fn task_with_data_connection(state: State, user_id: Uuid, otp: String) -> TaskResult<()> {
         let db: &PgPool = &state.db;
        Ok(())
}
```

How I achieved this by modifying the celery and task structs. 
When celery is being provided you have to provide a state object 

```
pub struct State<T> {
       db: &PgPool
}
// is initalized with
let state = Arc::new(Mutex::new(State::new(db.clone)))
// this can be passed to celery! macro when creating and tasks will have access to it.
```
I am modifying and injecting state into SendTask.  I took inspiration from the actix app data to do this since I am already using actix in my app. I think this approach might not be best way to achieve this.

**Comment by ehiggs** [comment]
> I would be up for it as I am using a modified version of rusty celery where I can have a state, but I am fairly new to doing rust libraries, so I don't know if my approach is good.

Do you have a gist with the modifications?
